README.md
# Forgery

![CI Status](https://github.com/kraftjectory/forgery/workflows/CI/badge.svg)
[![Hex Version](https://img.shields.io/hexpm/v/forgery.svg)](https://hex.pm/packages/forgery)

Forgery is a slim yet extensible data generator in Elixir.

## Installation

```elixir
def deps() do
  [{:forgery, "~> 0.1"}]
end
```

## Overview

Full documentation can be found at [https://hexdocs.pm/forgery](https://hexdocs.pm/forgery).

Forgery provides a few simple APIs to work with:

```elixir
defmodule User do
  defstruct [:id, :name, :password]
end

defmodule MyFactory do
  use Forgery

  def make(:user, fields) do
    fields
    |> put_new_field(:id, lazy(make_unique_integer()))
    |> put_new_field(:name, &("user#" <> Integer.to_string(&1.id)))
    |> create_struct(User)
  end
end

iex> import MyFactory
iex>
iex> %User{} = make(:user)
iex> %User{id: 42, name: "user#42"} = make(:user, id: 42)
iex> [%User{}, %User{}] = make_many(:user, 2)
```

And just as simple as that!

## Ecto integration

Forgery was built with easy Ecto integration in mind, though not limiting to it.

For example you use Ecto and have `MyRepo`. You can add a function, says `insert!` and `insert_many!`, into the factory:

```elixir
defmodule MyFactory do
  def insert!(factory_name, fields \\ %{}) do
    factory_name
    |> make(fields)
    |> MyRepo.insert!()
  end

  def insert_many!(factory_name, amount, fields \\ %{}) when amount >= 1 do
    [%schema{} | _] = entities = make_many(factory_name, amount, fields)

    {^amount, persisted_entities} = MyRepo.insert_all(schema, entities, returning: true)

    persisted_entities
  end
end

user = insert!(:user)
users = insert_many!(:user, 10, %{password: "1234567890"})
```

## Licensing

This software is licensed under [the ISC license](LICENSE).


CHANGELOG.md
# Changelog

## v0.2.0

__Breaking changes:__

  * Changed `Forgery.put_new_field/3` to be a plain function that accepts anonymous functions.
  * Infroduced `Forgery.lazy/1` to make `Forgery.put_new_field/3` usage more streamlined.

## v0.1.0

Initial release.


test/test_helper.exs
defmodule MyUser do
  defstruct [:id, :username, :password]
end

defmodule MyFactory do
  use Forgery

  def make(:user, fields) do
    fields
    |> put_new_field(:id, make_unique_integer())
    |> put_new_field(:username, "user" <> to_string(make_unique_integer()))
    |> create_struct(MyUser)
  end
end

ExUnit.start()


test/forgery_test.exs
defmodule ForgeryTest do
  use ExUnit.Case, async: true

  defmodule User do
    defstruct [:id, :name, :password]
  end

  defmodule MyFactory do
    use Forgery

    def make(:user, fields) do
      fields
      |> put_new_field(:id, lazy(make_unique_integer()))
      |> put_new_field(:name, &("user#" <> Integer.to_string(&1.id)))
      |> put_new_field(:name, fn _ -> raise("unexpected") end)
      |> create_struct(User)
    end
  end

  doctest Forgery, import: true

  test "make/1 and make/2" do
    assert %User{
             id: id,
             name: name
           } = MyFactory.make(:user)

    assert is_integer(id)
    assert id > 0
    assert name == "user##{id}"

    assert %User{
             id: 100,
             name: "user#100"
           } = MyFactory.make(:user, id: 100)

    assert %User{name: "John"} = MyFactory.make(:user, name: "John")
  end

  test "make_many/4" do
    assert [user1, user2] = MyFactory.make_many(:user, 2, %{password: "123456"})
    assert user1.id != user2.id
    assert user1.name != user2.name

    assert user1.password == "123456"
    assert user2.password == "123456"
  end
end


test/fortory_test.exs
defmodule FortoryTest do
  use ExUnit.Case, async: true

  doctest Fortory

  defmodule User do
    defstruct [:id, :username, :password]
  end

  defmodule DummyFactory do
    use Fortory

    def make(:user, fields) do
      id = make_unique_integer()

      make_foo = fn -> raise("asdfasdf") end

      fields
      |> put_new_field(:id, id)
      |> put_new_field(:username, "user#{id}")
      |> put_new_field(:username, make_foo.())
      |> create_struct(User)
    end
  end

  test "make/1 and make/2" do
    assert %User{
             id: id,
             username: username
           } = DummyFactory.make(:user)

    assert is_integer(id)
    assert id > 0
    assert username == "user#{id}"

    assert %User{
             id: 100,
             username: username
           } = DummyFactory.make(:user, id: 100)

    assert "user" <> _ = username

    assert %User{username: "john"} = DummyFactory.make(:user, username: "john")
  end

  test "make_many/4" do
    assert [user1, user2] = DummyFactory.make_many(:user, 2, %{password: "123456"})
    assert user1.id != user2.id
    assert user1.username != user2.username

    assert user1.password == "123456"
    assert user2.password == "123456"
  end
end


mix.exs
defmodule Forgery.MixProject do
  use Mix.Project

  @version "0.2.0"
  @source_url "https://github.com/craftjectory/forgery"

  def project() do
    [
      app: :forgery,
      version: @version,
      elixir: "~> 1.6",
      deps: deps(),

      # Hex.
      package: package(),
      description: description(),

      # Docs.
      name: "Forgery",
      docs: docs()
    ]
  end

  def application(), do: []

  defp deps() do
    [
      {:ex_doc, "~> 0.21", only: :dev, runtime: false}
    ]
  end

  defp description() do
    "A slim test data generator that does not compromise extensibility."
  end

  defp package() do
    [
      maintainers: ["Aleksei Magusev", "Cẩm Huỳnh"],
      licenses: ["ISC"],
      links: %{"GitHub" => @source_url}
    ]
  end

  defp docs() do
    [
      main: "Forgery",
      source_ref: "v#{@version}",
      source_url: @source_url,
      extras: [
        "README.md",
        "CHANGELOG.md"
      ]
    ]
  end
end


lib/fortory.ex
defmodule Fortory do
  @moduledoc """
  Fortory is a slim though extensible test data generator in Elixir.

  Fortory provides a few simple APIs to work with. To get started, you
  need to implement the `make/2` callback:

      defmodule MyUser do
        defstruct [:id, :username, :password]
      end

      defmodule MyFactory do
        use Fortory

        def make(:user, fields) do
          fields
          |> put_new_field(:id, make_unique_integer())
          |> put_new_field(:username, "user" <> to_string(make_unique_integer()))
          |> create_struct(MyUser)
        end
      end

      iex> import MyFactory
      iex>
      iex> %MyUser{} = make(:user)
      iex> %MyUser{id: 42} = make(:user, id: 42)
      iex> [%MyUser{}, %MyUser{}] = make_many(:user, 2)

  And just as simple as that!

  ## Ecto integration

  Fortory was built with easy Ecto integration in mind, though not limiting to it.

  For example if you use Ecto and have `MyRepo`. You can add a function, says `insert!`, into the factory:

      defmodule MyFactory do
        def insert!(factory, fields \\ %{}) do
          :user
          |> make(fields)
          |> MyRepo.insert!()
        end

        def insert_many!(factory, amount, fields \\ %{}) when amount >= 1 do
          [%schema{} | _] = entities = make_many(:user, amount, fields)

          {_, persisted_entities} = MyRepo.insert_all(schema, entities, returning: true)

          persisted_entities
        end
      end

      user = insert!(:user)
      users = insert_many!(:user, 10, %{password: "1234567890"})

  """

  @type factory() :: atom()

  @doc """
  Makes data from the given factory.

  The implementation of this callback should take in the factory name, as well and `fields`.


  """

  @callback make(factory :: factory(), fields :: Enumerable.t()) :: any()

  @doc """
  Make multiple data from the given factory.

  This function is roughly equivalent to:

      Enum.map(1..amount, fn _ -> make(factory) end)

  ### Example

      make_many(:users, 3)
      [
        %MyUser{id: 3, password: nil, username: "user3"},
        %MyUser{id: 5, password: nil, username: "user4"},
        %MyUser{id: 7, password: nil, username: "user5"},
      ]


  """
  @callback make_many(factory :: factory(), amount :: integer(), fields :: Enumerable.t()) ::
              list(any())

  defmacro __using__(_) do
    quote location: :keep do
      import Fortory

      @behaviour Fortory

      def make(factory, fields \\ %{})

      def make_many(factory, amount, fields \\ %{}) when is_integer(amount) do
        for _ <- 1..amount, do: make(factory, fields)
      end
    end
  end

  @doc """
  Lazily evaluate and put `lazy_value` into `name` if `name` does not exist in `fields`.

      iex> import Fortory
      iex>
      iex> fields = %{foo: 1}
      iex> put_new_field(fields, :foo, 100 + 2)
      %{foo: 1}
      iex> put_new_field(fields, :bar, 100)
      %{foo: 1, bar: 100}

  Note that `lazy_value` is only evaluated when it is needed. For instance, in the
  following example, `make_foo.()` will not be invoked.

      iex> import Fortory
      iex>
      iex> make_foo = fn -> raise("I am invoked") end
      iex> fields = %{foo: 1}
      iex> put_new_field(fields, :foo, make_foo.())
      %{foo: 1}
  """

  @spec put_new_field(fields :: Enumerable.t(), name :: any(), lazy_value :: any()) :: map()
  defmacro put_new_field(fields, name, lazy_value) do
    quote do
      unquote(fields)
      |> Map.new()
      |> Map.put_new_lazy(unquote(name), fn -> unquote(lazy_value) end)
    end
  end

  @doc """
  Create struct of `module` from `fields`.

      iex> import Fortory
      iex>
      iex> create_struct(%{id: 1, username: "john", password: "123456"}, MyUser)
      %MyUser{id: 1, password: "123456", username: "john"}

  """

  @spec create_struct(fields :: Enumerable.t(), module :: atom()) :: struct()
  def create_struct(fields, module) do
    struct!(module, fields)
  end

  @doc """
  Returns monotonically increasing unique integer. It would be useful when it comes to
  generate unique serial IDs.
  """
  @spec make_unique_integer() :: pos_integer()
  def make_unique_integer() do
    System.unique_integer([:monotonic, :positive])
  end
end


lib/forgery.ex
defmodule Forgery do
  @moduledoc """
  Forgery is a slim yet extensible data generator in Elixir.

  Forgery provides a few simple APIs to work with. To get started, you
  need to implement the `make/2` callback:

      defmodule User do
        defstruct [:id, :name, :password]
      end

      defmodule MyFactory do
        use Forgery

        def make(:user, fields) do
          fields
          |> put_new_field(:id, lazy(make_unique_integer()))
          |> put_new_field(:name, &("user#" <> Integer.to_string(&1.id)))
          |> create_struct(User)
        end
      end

      iex> import MyFactory
      iex>
      iex> %User{} = make(:user)
      iex> %User{id: 42, name: "user#42"} = make(:user, id: 42)
      iex> [%User{}, %User{}] = make_many(:user, 2)

  And just as simple as that!

  ## Ecto integration

  Forgery was built with easy Ecto integration in mind, though not limiting to it.

  For example if you use Ecto and have `MyRepo`. You can add a function, says `insert!`, into the factory:

      defmodule MyFactory do
        def insert!(factory_name, fields \\ %{}) do
          factory_name
          |> make(fields)
          |> MyRepo.insert!()
        end

        def insert_many!(factory_name, amount, fields \\ %{}) when amount >= 1 do
          [%schema{} | _] = entities = make_many(factory_name, amount, fields)

          {_, persisted_entities} = MyRepo.insert_all(schema, entities, returning: true)

          persisted_entities
        end
      end

      user = insert!(:user)
      users = insert_many!(:user, 10, %{password: "1234567890"})

  """

  @type factory_name() :: atom()

  @doc """
  Makes data from the given factory.

  The implementation of this callback should take in the factory name, as well and `fields`.
  """

  @callback make(factory_name(), fields :: Enumerable.t()) :: any()

  @doc """
  Make multiple data from the given factory.

  This function is roughly equivalent to:

      Enum.map(1..amount, fn _ -> make(factory_name) end)

  ### Example

      make_many(:users, 3)
      [
        %User{id: 3, password: nil, name: "user#3"},
        %User{id: 5, password: nil, name: "user#5"},
        %User{id: 7, password: nil, name: "user#7"}
      ]

  """
  @callback make_many(factory_name(), amount :: integer(), fields :: Enumerable.t()) ::
              list(any())

  defmacro __using__(_) do
    quote location: :keep do
      import Forgery

      @behaviour Forgery

      def make(factory_name, fields \\ %{})

      def make_many(factory_name, amount, fields \\ %{}) when is_integer(amount) do
        for _ <- 1..amount, do: make(factory_name, fields)
      end
    end
  end

  @doc """
  Lazily evaluates `value_setter` and puts the result into `key` if it does not exist in `fields`.

  The `value_setter` function receives `fields` as an argument.

      iex> make_foo = fn _ -> raise("I am invoked") end
      iex> fields = %{foo: 1}
      iex> put_new_field(fields, :foo, make_foo)
      %{foo: 1}
      iex> put_new_field(fields, :bar, &(&1.foo + 100))
      %{foo: 1, bar: 101}

  There is also helper macro `lazy/1`:

      iex> fields = %{foo: 2}
      iex> put_new_field(fields, :foo, lazy(10 * 10))
      %{foo: 2}

  """
  @spec put_new_field(
          fields :: Enumerable.t(),
          key :: any(),
          value_setter :: (fields :: map() -> any())
        ) :: map()
  def put_new_field(fields, key, value_setter) when is_function(value_setter, 1) do
    case Map.new(fields) do
      %{^key => _value} = fields ->
        fields

      fields ->
        Map.put(fields, key, value_setter.(fields))
    end
  end

  @doc """
  Wraps the given `expr` into an anonymous function.

  It is equivalent to `fn _ -> expr end`.
  """
  defmacro lazy(expr) do
    quote do
      fn _ -> unquote(expr) end
    end
  end

  @doc """
  Create struct of `module` from `fields`.

  See `Kernel.struct!/2` for more information.

      iex> create_struct(%{id: 1, name: "John", password: "123456"}, User)
      %User{id: 1, password: "123456", name: "John"}

  """

  @spec create_struct(fields :: Enumerable.t(), module() | struct()) :: struct()
  def create_struct(fields, module) do
    struct!(module, fields)
  end

  @doc """
  Returns monotonically increasing unique integer. It would be useful when it comes to
  generate unique serial IDs.
  """
  @spec make_unique_integer() :: pos_integer()
  def make_unique_integer() do
    System.unique_integer([:monotonic, :positive])
  end
end
